modules/core/shared/src/main/scala/io/circe/JsonNumber.scala

```diff
@@ -236,6 +236,18 @@ private[circe] final case class JsonFloat(value: Float) extends JsonNumber {
  */
 object JsonNumber {
 
+  /**
+   * Constant pool of integer numbers between -128 to 127, similar to the JVM's constant pool for integers.
+   */
+  private[this] val jsonLongConstantPool: Array[JsonLong] =
+    Array.tabulate(256)(x => JsonLong((x - 128).toLong))
+
+  private[circe] final def fromLong(value: Long): JsonLong = {
+    val idx = value + 128
+    if (idx >= 0 && idx < jsonLongConstantPool.length) jsonLongConstantPool.apply(idx.toInt)
+    else JsonLong(value)
+  }
```

Comment:
I think it would make more sense to store `JNumber` since thats what the end user will see the most, and it'll save 1 additional allocation.

```suggestion
  private[this] val jsonLongConstantPool: Array[JNumber] =
    Array.tabulate(256)(x => JsonLong((x - 128).toLong))

  private[circe] final def fromLong(value: Long): JNumber = {
    val idx = value + 128
    if (idx >= 0 && idx < jsonLongConstantPool.length) jsonLongConstantPool.apply(idx.toInt)
    else JNumber(value)
  }
```